synapse/rest/client/v1/room.py

```diff
@@ -266,6 +267,254 @@ def on_PUT(self, request, room_id, event_type, txn_id):
+        if chunk_id_from_query:
+            last_event_in_chunk = events_to_create[len(events_to_create) - 1]
+            last_event_in_chunk["content"][
+                EventContentFields.MSC2716_CHUNK_ID
+            ] = chunk_id_from_query
+
+        # Add an "insertion" event to the start of each chunk (next to the oldest
+        # event in the chunk) so the next chunk can be connected to this one.
+        next_chunk_id = random_string(64)
+        insertion_event = {
+            "type": EventTypes.MSC2716_INSERTION,
+            "sender": requester.user.to_string(),
+            "content": {
+                EventContentFields.MSC2716_NEXT_CHUNK_ID: next_chunk_id,
+                EventContentFields.MSC2716_HISTORICAL: True,
+                # TODO: Why is `body` necessary for this to show up in /messages
+                "body": "TODO_REMOVE - INSERTION",
+            },
+            # Since the insertion event is put at the end of the chunk,
+            # copy the origin_server_ts from the last event we're inserting
+            "origin_server_ts": events_to_create[0]["origin_server_ts"],
+        }
+        # Prepend the insertion event to the start of the chunk
+        events_to_create = [insertion_event] + events_to_create
+
+        inherited_depth = await self.inherit_depth_from_prev_ids(prev_events_from_query)
+        logger.info("inherited_depth %s", inherited_depth)
+
+        event_ids = []
+        prev_event_ids = prev_events_from_query
+        events_to_persist = []
+        for ev in events_to_create:
+            assert_params_in_dict(ev, ["type", "origin_server_ts", "content", "sender"])
+
+            # Mark all events as historical
+            # This has important semantics within the Synapse internals to backfill properly
+            ev["content"][EventContentFields.MSC2716_HISTORICAL] = True
+
+            event_dict = {
+                "type": ev["type"],
+                "origin_server_ts": ev["origin_server_ts"],
+                "content": ev["content"],
+                "room_id": room_id,
+                "sender": ev["sender"],  # requester.user.to_string(),
+                "prev_events": prev_event_ids.copy(),
+            }
+
+            event, context = await self.event_creation_handler.create_event(
+                requester,
+                event_dict,
+                prev_event_ids=event_dict.get("prev_events"),
+                # TODO: Do we need to use `self.auth.compute_auth_events(...)` to filter the `auth_event_ids`?
+                auth_event_ids=auth_event_ids,
+                depth=inherited_depth,
+            )
+            current_state_ids = await context.get_current_state_ids()
+            logger.info(
+                "bulksend event=%s current_state_ids=%s",
+                event,
+                current_state_ids,
+            )
+
+            # TODO: Should we add the same `hs.is_mine_id(event.sender)` assert check that `create_and_send_nonmember_event` has?
```

Comment:
I think we should, just to avoid any confusion if an appservice tries to do something silly.